src/REP15.sol

```diff
@@ -0,0 +1,466 @@
+
+    if (!delegation.isPending()) revert REP15NonexistentPendingOwnershipDelegation(tokenId);
+
+    return (delegation.delegatee, delegation.until);
+  }
+
+  /**
+   * @dev Ensures the context is existent and returns the controller of the context `ctxHash`.
+   */
+  function _requireControlled(bytes32 ctxHash) internal view returns (address controller) {
+    controller = _contexts[ctxHash].controller;
+    if (controller == address(0)) revert REP15NonexistentContext(ctxHash);
+  }
+
+  /**
+   * @dev Checks if the context is active and `controller` is the controller of the context `ctxHash`.
+   */
+  function _checkAuthorizedController(address controller, bytes32 ctxHash) internal view virtual {
+    if (controller != _requireControlled(ctxHash)) revert REP15InvalidController(controller);
+  }
+
+  /**
+   * @dev Internal function to create or update a context.
+   *
+   * The `auth` argument is optional. If the value passed is non 0, then this function will check that
+   * `auth` is the controller of `ctxHash` before updating the context.
+   *
+   * Emits a {ContextUpdated} event.
+   */
+  function _updateContext(bytes32 ctxHash, address controller, uint64 detachingDuration, address auth) internal virtual {
+    if (controller == address(0)) revert REP15InvalidController(address(0));
+    if (detachingDuration > maxDetachingDuration()) revert REP15ExceededMaxDetachingDuration(detachingDuration);
+
+    REP15Utils.Context storage context = _contexts[ctxHash];
+
+    if (auth != address(0)) {
+      // Updating context
+      _checkAuthorizedController(auth, ctxHash);
+    } else {
+      // Creating context
+      if (context.controller != address(0)) revert REP15ExistentContext(ctxHash);
+    }
+
+    context.controller = controller;
+    context.detachingDuration = detachingDuration;
+
+    emit ContextUpdated(ctxHash, controller, detachingDuration);
+  }
+
+  /**
+   * @dev Ensures the context is attached to the token and returns the token context storage pointer.
+   * If `checkNotRequestedForDetachment` is true, this function will check that the context has not requested for detachment.
+   */
+  function _requireAttachedTokenContext(bytes32 ctxHash, uint256 tokenId, bool checkNotRequestedForDetachment)
+    internal
+    view
+    returns (REP15Utils.TokenContext storage tokenContext)
+  {
+    tokenContext = _tokenContext[tokenId][ctxHash];
+
+    if (!tokenContext.attached) revert REP15NonexistentAttachedContext(ctxHash, tokenId);
+
+    if (checkNotRequestedForDetachment && tokenContext.hasRequestedForDetachment()) {
+      revert REP15RequestedForDetachment(ctxHash, tokenId);
+    }
+  }
+
+  /**
+   * @dev Ensures the context is not attached to the token and returns the token context storage pointer.
+   */
+  function _requireNotAttachedTokenContext(bytes32 ctxHash, uint256 tokenId)
+    internal
+    view
+    returns (REP15Utils.TokenContext storage tokenContext)
```

Comment:
Should not load storage pointer inside view function
```suggestion
    returns (REP15Utils.TokenContext memory tokenContext)
```

Reply:
Your code cannot modify the state and it costs more gas.

Reply:
I have seen that Uniswap also uses this pattern in [here](https://github.com/Uniswap/v3-core/blob/v1.0.0/contracts/UniswapV3Pool.sol#L310). Please feel free to use it.